// @vue-storefront/shopify-api/server/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@vue-storefront/core');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function getProduct(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var chosenVariant_1, getProductByHandleQuery, getrelatedProductsByIdQuery;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!params.slug) return [3 /*break*/, 1];
                    chosenVariant_1 = [];
                    if (params.selectedOptions && Object.keys(params.selectedOptions).length > 0) {
                        chosenVariant_1 = Object.entries(params.selectedOptions).map(function (k) { return ({ name: k[0], value: k[1] }); });
                    }
                    getProductByHandleQuery = context.client.graphQLClient.query(function (root) {
                        root.add('productByHandle', { args: { handle: params.slug } }, function (productByHandle) {
                            // get product basic information
                            productByHandle.add('id');
                            productByHandle.add('title');
                            productByHandle.add('description');
                            productByHandle.add('descriptionHtml');
                            productByHandle.add('handle');
                            productByHandle.add('tags');
                            productByHandle.add('availableForSale');
                            productByHandle.add('totalInventory');
                            productByHandle.add('vendor');
                            productByHandle.add('variantBySelectedOptions', { args: { selectedOptions: chosenVariant_1 } }, function (selectedVariant) {
                                selectedVariant.add('id');
                                selectedVariant.add('title');
                                selectedVariant.add('sku');
                                selectedVariant.add('availableForSale');
                                selectedVariant.add('quantityAvailable');
                                selectedVariant.addField('image', { args: {} }, function (image) {
                                    image.add('altText');
                                    image.add('originalSrc');
                                    image.add('transformedSrc');
                                });
                                selectedVariant.addField('priceV2', function (price) {
                                    price.add('amount');
                                    price.add('currencyCode');
                                });
                                selectedVariant.addField('compareAtPriceV2', function (price) {
                                    price.add('amount');
                                    price.add('currencyCode');
                                });
                            });
                            productByHandle.add('options', {}, function (options) {
                                options.add('name');
                                options.add('values');
                            });
                            productByHandle.addConnection('collections', { args: { first: 20 } }, function (collection) {
                                collection.add('title');
                                collection.add('handle');
                            });
                            productByHandle.addConnection('images', { args: { first: 20 } }, function (image) {
                                image.add('id');
                                image.add('altText');
                                image.add('originalSrc');
                                image.add('transformedSrc');
                            });
                            productByHandle.addConnection('variants', { args: { first: 1 } }, function (variants) {
                                variants.add('title');
                                variants.add('price');
                                variants.add('weight');
                                variants.add('available');
                                variants.add('sku');
                                variants.add('compareAtPrice');
                                variants.addField('image', { args: {} }, function (image) {
                                    image.add('id');
                                    image.add('altText');
                                    image.add('originalSrc');
                                    image.add('transformedSrc');
                                });
                                variants.addField('selectedOptions', {}, function (selectedOptions) {
                                    selectedOptions.add('name');
                                    selectedOptions.add('value');
                                });
                                variants.addField('product', {}, function (product) {
                                    product.add('id');
                                    product.add('title');
                                    product.add('availableForSale');
                                    product.add('handle');
                                    product.add('description');
                                    product.add('descriptionHtml');
                                    product.addConnection('images', { args: { first: 20 } }, function (images) {
                                        images.add('id');
                                        images.add('altText');
                                        images.add('originalSrc');
                                        images.add('transformedSrc');
                                    });
                                    product.add('productType');
                                    product.addField('options', {}, function (options) {
                                        options.add('name');
                                        options.add('values');
                                    });
                                });
                            });
                        });
                    });
                    return [2 /*return*/, context.client.graphQLClient
                            .send(getProductByHandleQuery)
                            .then(function (_a) {
                            var model = _a.model, product = _a.product;
                            if (model) {
                                return model.productByHandle;
                            }
                        })];
                case 1:
                    if (!params.related) return [3 /*break*/, 2];
                    getrelatedProductsByIdQuery = context.client.graphQLClient.query(function (root) {
                        root.add('productRecommendations', { args: { productId: params.productId } }, function (relatedProductsById) {
                            // get product basic information
                            relatedProductsById.add('id');
                            relatedProductsById.add('title');
                            relatedProductsById.add('handle');
                            relatedProductsById.add('options', {}, function (options) {
                                options.add('name');
                                options.add('values');
                            });
                            relatedProductsById.addConnection('images', { args: { first: 20 } }, function (image) {
                                image.add('id');
                                image.add('altText');
                                image.add('originalSrc');
                                image.add('transformedSrc');
                            });
                            relatedProductsById.addConnection('collections', { args: { first: 20 } }, function (collection) {
                                collection.add('title');
                                collection.add('handle');
                            });
                            relatedProductsById.addConnection('variants', { args: { first: 20 } }, function (variants) {
                                variants.add('title');
                                variants.add('price');
                                variants.add('weight');
                                variants.add('available');
                                variants.add('sku');
                                variants.add('compareAtPrice');
                                variants.addField('image', { args: {} }, function (image) {
                                    image.add('id');
                                    image.add('altText');
                                    image.add('originalSrc');
                                    image.add('transformedSrc');
                                });
                                variants.addField('selectedOptions', {}, function (selectedOptions) {
                                    selectedOptions.add('name');
                                    selectedOptions.add('value');
                                });
                                variants.addField('product', {}, function (product) {
                                    product.add('id');
                                    product.add('title');
                                    product.add('availableForSale');
                                    product.add('handle');
                                    product.add('description');
                                    product.add('descriptionHtml');
                                    product.addConnection('images', { args: { first: 20 } }, function (images) {
                                        images.add('id');
                                        images.add('altText');
                                        images.add('originalSrc');
                                        images.add('transformedSrc');
                                    });
                                    product.add('productType');
                                    product.addField('options', {}, function (options) {
                                        options.add('name');
                                        options.add('values');
                                    });
                                });
                            });
                        });
                    });
                    return [2 /*return*/, context.client.graphQLClient
                            .send(getrelatedProductsByIdQuery)
                            .then(function (_a) {
                            var model = _a.model;
                            if (model) {
                                return model.productRecommendations;
                            }
                        })];
                case 2:
                    if (!params.id) return [3 /*break*/, 4];
                    return [4 /*yield*/, context.client.product.fetch(params.id).then(function (product) {
                            return product;
                        })];
                case 3: return [2 /*return*/, _a.sent()];
                case 4: return [4 /*yield*/, context.client.product
                        .fetchQuery({ first: (params.limit ? params.limit : 20), sortKey: (params.sortBy ? params.sortBy : 'CREATED_AT'), reverse: false })
                        .then(function (products) {
                        return products;
                    })];
                case 5: return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function getCategory(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            // Use the built-in function
            if (params.slug !== '') {
                return [2 /*return*/, context.client.collection.fetchByHandle(params.slug).then(function (collection) {
                        // Collection with all default fields
                        return collection;
                    })];
            }
            else {
                return [2 /*return*/, context.client.collection.fetchAll().then(function (collection) {
                        // Collection with all default fields
                        return collection;
                    })];
            }
        });
    });
}

/* TODO: Fetch custom client directly, may be using context  */
var changePasswordMutation = function (context) {
    var customerAccessToken = context.client.graphQLClient.variable('customerAccessToken', 'String!');
    var customer = context.client.graphQLClient.variable('customer', 'CustomerUpdateInput!');
    return context.client.graphQLClient.mutation('customerUpdate', [customerAccessToken, customer], function (root) {
        root.add('customerUpdate', { args: { customerAccessToken: customerAccessToken, customer: customer } }, function (customer) {
            customer.add('customer', function (fields) {
                fields.add('id');
                fields.add('displayName');
                fields.add('email');
                fields.add('firstName');
                fields.add('lastName');
                fields.add('phone');
            });
            customer.add('customerAccessToken', function (token) {
                token.add('accessToken');
                token.add('expiresAt');
            });
            customer.add('customerUserErrors', function (error) {
                error.add('code');
                error.add('field');
                error.add('message');
            });
        });
    });
};
var forgotPasswordMutation = function (context) {
    var email = context.client.graphQLClient.variable('email', 'String!');
    return context.client.graphQLClient.mutation('customerRecover', [email], function (root) {
        root.add('customerRecover', { args: { email: email } }, function (customer) {
            customer.add('customerUserErrors', function (error) {
                error.add('code');
                error.add('field');
                error.add('message');
            });
        });
    });
};
var customerAddressDeleteMutation = function (context) {
    var id = context.client.graphQLClient.variable('id', 'ID!');
    var customerAccessToken = context.client.graphQLClient.variable('customerAccessToken', 'String!');
    return context.client.graphQLClient.mutation('customerAddressDelete', [id, customerAccessToken], function (root) {
        root.add('customerAddressDelete', { args: { id: id, customerAccessToken: customerAccessToken } }, function (customer) {
            customer.add('customerUserErrors', function (error) {
                error.add('code');
                error.add('field');
                error.add('message');
            });
            customer.add('deletedCustomerAddressId');
        });
    });
};
var customerAddressAddMutation = function (context) {
    var address = context.client.graphQLClient.variable('address', 'MailingAddressInput!');
    var customerAccessToken = context.client.graphQLClient.variable('customerAccessToken', 'String!');
    return context.client.graphQLClient.mutation('customerAddressCreate', [customerAccessToken, address], function (root) {
        root.add('customerAddressCreate', { args: { address: address, customerAccessToken: customerAccessToken } }, function (customer) {
            customer.add('customerAddress', function (addressInfo) {
                addressInfo.add('id');
            });
            customer.add('customerUserErrors', function (error) {
                error.add('code');
                error.add('field');
                error.add('message');
            });
        });
    });
};
var customerAddressUpdateMutation = function (context) {
    var customerAccessToken = context.client.graphQLClient.variable('customerAccessToken', 'String!');
    var id = context.client.graphQLClient.variable('id', 'ID!');
    var address = context.client.graphQLClient.variable('address', 'MailingAddressInput!');
    return context.client.graphQLClient.mutation('customerAddressUpdate', [customerAccessToken, id, address], function (root) {
        root.add('customerAddressUpdate', { args: { address: address, id: id, customerAccessToken: customerAccessToken } }, function (customer) {
            customer.add('customerAddress', function (addressInfo) {
                addressInfo.add('id');
            });
            customer.add('customerUserErrors', function (error) {
                error.add('code');
                error.add('field');
                error.add('message');
            });
        });
    });
};
var editProfileMutation = function (context) {
    var customerAccessToken = context.client.graphQLClient.variable('customerAccessToken', 'String!');
    var customer = context.client.graphQLClient.variable('customer', 'CustomerUpdateInput!');
    return context.client.graphQLClient.mutation('customerUpdate', [customerAccessToken, customer], function (root) {
        root.add('customerUpdate', { args: { customerAccessToken: customerAccessToken, customer: customer } }, function (customer) {
            customer.add('customer', function (fields) {
                fields.add('id');
                fields.add('displayName');
                fields.add('email');
                fields.add('firstName');
                fields.add('lastName');
                fields.add('phone');
            });
            customer.add('customerAccessToken', function (token) {
                token.add('accessToken');
                token.add('expiresAt');
            });
            customer.add('customerUserErrors', function (error) {
                error.add('code');
                error.add('field');
                error.add('message');
            });
        });
    });
};
var signInMutation = function (context) {
    var input = context.client.graphQLClient.variable('input', 'CustomerAccessTokenCreateInput!');
    return context.client.graphQLClient.mutation('customerAccessTokenCreate', [input], function (root) {
        root.add('customerAccessTokenCreate', { args: { input: input } }, function (customer) {
            customer.add('customerAccessToken', function (token) {
                token.add('accessToken');
                token.add('expiresAt');
            });
            customer.add('customerUserErrors', function (error) {
                error.add('code');
                error.add('field');
                error.add('message');
            });
        });
    });
};
var signOutMutation = function (context) {
    var customerAccessToken = context.client.graphQLClient.variable('customerAccessToken', 'String!');
    return context.client.graphQLClient.mutation('customerAccessTokenDelete', [customerAccessToken], function (root) {
        root.add('customerAccessTokenDelete', { args: { customerAccessToken: customerAccessToken } }, function (customer) {
            customer.add('deletedAccessToken');
            customer.add('deletedCustomerAccessTokenId');
            customer.add('userErrors', function (error) {
                error.add('field');
                error.add('message');
            });
        });
    });
};
var signUpMutation = function (context) {
    var input = context.client.graphQLClient.variable('input', 'CustomerCreateInput!');
    return context.client.graphQLClient.mutation('customerCreate', [input], function (root) {
        root.add('customerCreate', { args: { input: input } }, function (customer) {
            customer.add('customer', function (token) {
                token.add('id');
            });
            customer.add('customerUserErrors', function (error) {
                error.add('code');
                error.add('field');
                error.add('message');
            });
        });
    });
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function editProfile(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var token, profile, data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    token = params.token, profile = params.profile;
                    data = {
                        customerAccessToken: token,
                        customer: profile
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(editProfileMutation(context), data).then(function (_a) {
                            var model = _a.model;
                            return model;
                        })];
                case 1: 
                // send user data to authenticate, return token if valid
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function getBlogPosts(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var articlesQuery;
        return __generator(this, function (_a) {
            articlesQuery = context.client.graphQLClient.query(function (root) {
                root.addConnection('articles', { args: { first: 20 } }, function (article) {
                    article.add('title');
                    article.add('handle');
                    article.add('url');
                    article.add('excerpt');
                    article.add('publishedAt');
                    article.addField('image', {}, function (image) {
                        image.add('id');
                        image.add('altText');
                        image.add('originalSrc');
                    });
                    article.addField('authorV2', {}, function (author) {
                        author.add('name');
                        author.add('email');
                    });
                });
            });
            // Call the send method with the custom query
            context.client.graphQLClient.send(articlesQuery).then(function (_a) {
                var model = _a.model, articles = _a.articles;
                if (model) {
                    return model.articles;
                }
            });
            return [2 /*return*/];
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function getPages(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var pageByHandleQuery, pagesQuery;
        return __generator(this, function (_a) {
            // Use the built-in function
            if (params.slug) {
                pageByHandleQuery = context.client.graphQLClient.query(function (root) {
                    root.add('pageByHandle', { args: { handle: params.slug } }, function (pageByHandle) {
                        pageByHandle.add('id');
                        pageByHandle.add('title');
                        pageByHandle.add('handle');
                        pageByHandle.add('url');
                        pageByHandle.add('bodySummary');
                        pageByHandle.add('body');
                        pageByHandle.add('createdAt');
                    });
                });
                // Call the send method with the custom query
                return [2 /*return*/, context.client.graphQLClient.send(pageByHandleQuery).then(function (_a) {
                        var model = _a.model, content = _a.content;
                        return model.pageByHandle;
                    })];
            }
            else {
                pagesQuery = context.client.graphQLClient.query(function (root) {
                    root.addConnection('pages', { args: { first: params.limit ? params.limit : 20 } }, function (article) {
                        article.add('id');
                        article.add('title');
                        article.add('handle');
                        article.add('url');
                        article.add('bodySummary');
                        article.add('body');
                        article.add('createdAt');
                    });
                });
                // Call the send method with the custom query
                return [2 /*return*/, context.client.graphQLClient.send(pagesQuery).then(function (_a) {
                        var model = _a.model, content = _a.content;
                        if (model) {
                            return model;
                        }
                    })];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function updateCart(context, params, _customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var currentCart, product, quantity, checkoutID, itemID;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    currentCart = params.currentCart, product = params.product, quantity = params.quantity;
                    checkoutID = currentCart.id;
                    itemID = product.id;
                    return [4 /*yield*/, context.client.checkout.updateLineItems(checkoutID, { id: product.id, quantity: quantity }).then(function (checkout) { return checkout; })];
                case 1: 
                // Add an item to the checkout
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function createCart(context, _params, _customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, context.client.checkout.create().then(function (checkout) {
                        // return checkOut ID
                        return checkout.id;
                    })];
                case 1: 
                // initiate the cart
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function addToCart(context, params, _customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var currentCart, product, quantity, checkoutID, lineItemsToAdd;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    currentCart = params.currentCart, product = params.product, quantity = params.quantity;
                    checkoutID = currentCart.id;
                    lineItemsToAdd = [{
                            variantId: product.variantBySelectedOptions && product.variantBySelectedOptions !== null ? product.variantBySelectedOptions.id : product.variantId,
                            quantity: quantity
                        }];
                    return [4 /*yield*/, context.client.checkout.addLineItems(checkoutID, lineItemsToAdd).then(function (checkout) { return checkout; })];
                case 1: 
                // Add an item to the checkout
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function removeFromCart(context, params, _customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var currentCart, product, checkoutID, lineItemIdsToRemove;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    currentCart = params.currentCart, product = params.product;
                    checkoutID = currentCart.id;
                    lineItemIdsToRemove = [
                        product.id
                    ];
                    return [4 /*yield*/, context.client.checkout.removeLineItems(checkoutID, lineItemIdsToRemove).then(function (checkout) { return checkout; })];
                case 1: 
                // Add an item to the checkout
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function signUp(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    data = {
                        input: params
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(signUpMutation(context), data).then(function (_a) {
                            var model = _a.model;
                            return model;
                        })];
                case 1: 
                // send userdata to register
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function signIn(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var username, password, data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    username = params.username, password = params.password;
                    data = {
                        input: {
                            email: username,
                            password: password
                        }
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(signInMutation(context), data).then(function (_a) {
                            var model = _a.model;
                            return model;
                        })];
                case 1: 
                // send user data to authenticate, return token if valid
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function signOut(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    data = {
                        customerAccessToken: params
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(signOutMutation(context), data).then(function (_a) {
                            var model = _a.model;
                            return model;
                        })];
                case 1: 
                // Remove customer access token
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function changePassword(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var token, newPassword, data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    token = params.token, newPassword = params.newPassword;
                    data = {
                        customerAccessToken: token,
                        customer: {
                            password: newPassword
                        }
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(changePasswordMutation(context), data).then(function (_a) {
                            var model = _a.model;
                            return model;
                        })];
                case 1: 
                // Remove customer access token
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function forgotPassword(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var username, data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    username = params.username;
                    data = {
                        email: username
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(forgotPasswordMutation(context), data)];
                case 1: 
                // Remove customer access token
                return [2 /*return*/, _a.sent()];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function fetchCustomer(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var getCustomeInfo;
        return __generator(this, function (_a) {
            getCustomeInfo = context.client.graphQLClient.query(function (root) {
                root.add('customer', { args: { customerAccessToken: params } }, function (customerInfo) {
                    customerInfo.add('id');
                    customerInfo.add('displayName');
                    customerInfo.add('email');
                    customerInfo.add('firstName');
                    customerInfo.add('lastName');
                    customerInfo.add('phone');
                    customerInfo.add('tags');
                    customerInfo.add('acceptsMarketing');
                });
            });
            // send user data to authenticate, return token if valid
            return [2 /*return*/, context.client.graphQLClient
                    .send(getCustomeInfo)
                    .then(function (_a) {
                    var model = _a.model, product = _a.product;
                    if (model) {
                        return model;
                    }
                })];
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function fetchOrders(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var getCustomerOrders;
        return __generator(this, function (_a) {
            getCustomerOrders = context.client.graphQLClient.query(function (root) {
                root.add('customer', { args: { customerAccessToken: params } }, function (customerOrder) {
                    customerOrder.addConnection('orders', { args: { first: 20 } }, function (order) {
                        order.add('id');
                        order.add('name');
                        order.add('orderNumber');
                        order.add('processedAt');
                        order.add('financialStatus');
                        order.add('fulfillmentStatus');
                        order.addField('currentTotalPrice', {}, function (curTotPrice) {
                            curTotPrice.add('amount');
                            curTotPrice.add('currencyCode');
                        });
                        order.addField('currentTotalTax', {}, function (curTotTax) {
                            curTotTax.add('amount');
                            curTotTax.add('currencyCode');
                        });
                        order.addField('currentSubtotalPrice', {}, function (curSubTotPrice) {
                            curSubTotPrice.add('amount');
                            curSubTotPrice.add('currencyCode');
                        });
                        order.addConnection('lineItems', { args: { first: 250 } }, function (Lineitem) {
                            Lineitem.add('title');
                            Lineitem.add('quantity');
                            Lineitem.addField('originalTotalPrice', {}, function (originaldPrice) {
                                originaldPrice.add('amount');
                                originaldPrice.add('currencyCode');
                            });
                            Lineitem.addField('variant', {}, function (variant) {
                                variant.add('sku');
                                variant.addField('product', {}, function (parentProduct) {
                                    parentProduct.add('id');
                                    parentProduct.add('handle');
                                });
                                variant.addField('image', {}, function (parentProduct) {
                                    parentProduct.add('altText');
                                    parentProduct.add('originalSrc');
                                });
                            });
                        });
                        order.addField('successfulFulfillments', {}, function (trackingInfo) {
                            trackingInfo.addField('trackingInfo', {}, function (tracking) {
                                tracking.add('number');
                                tracking.add('url');
                            });
                        });
                        order.addField('shippingAddress', {}, function (saddress) {
                            saddress.add('name');
                            saddress.add('phone');
                            saddress.add('formatted');
                        });
                        order.addField('totalShippingPriceV2', {}, function (TotalShippingPrice) {
                            TotalShippingPrice.add('amount');
                            TotalShippingPrice.add('currencyCode');
                        });
                    });
                });
            });
            return [2 /*return*/, context.client.graphQLClient
                    .send(getCustomerOrders)
                    .then(function (_a) {
                    var model = _a.model, product = _a.product;
                    if (model) {
                        return model;
                    }
                })];
        });
    });
}

/* eslint-disable func-names */
var addressesQuery = function (pages, token, context) {
    return context.client.graphQLClient.query(function (root) {
        root.add('customer', {
            args: {
                customerAccessToken: token
            }
        }, function (customer) {
            customer.add('id');
            customer.addConnection('addresses', { args: { first: 10 } }, function (address) {
                address.add('address1');
                address.add('address2');
                address.add('city');
                address.add('company');
                address.add('firstName');
                address.add('lastName');
                address.add('name');
                address.add('phone');
                address.add('province');
                address.add('provinceCode');
                address.add('zip');
                address.add('formattedArea');
            });
        });
    });
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function fetchAddresses(context, params, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        var addresses;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, context.client.graphQLClient.send(addressesQuery(10, params, context)).then(function (_a) {
                        var model = _a.model;
                        if (model) {
                            return model.customer;
                        }
                    })];
                case 1:
                    addresses = _a.sent();
                    return [2 /*return*/, addresses];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function deleteAddress(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var token, AddressId, data, deletedAddressId;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    token = params.token, AddressId = params.AddressId;
                    data = {
                        id: AddressId,
                        customerAccessToken: token
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(customerAddressDeleteMutation(context), data).then(function (_a) {
                            var model = _a.model;
                            if (model) {
                                return model.customerAddressDelete;
                            }
                        })];
                case 1:
                    deletedAddressId = _a.sent();
                    return [2 /*return*/, deletedAddressId];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function addAddress(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var token, address, data, addedAddressId;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    token = params.token, address = params.address;
                    data = {
                        address: address,
                        customerAccessToken: token
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(customerAddressAddMutation(context), data).then(function (_a) {
                            var model = _a.model;
                            if (model) {
                                return model.customerAddressCreate;
                            }
                        })];
                case 1:
                    addedAddressId = _a.sent();
                    return [2 /*return*/, addedAddressId];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function updateAddress(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var token, AddressId, address, data, updatedAddressId;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    token = params.token, AddressId = params.AddressId, address = params.address;
                    data = {
                        id: AddressId,
                        address: address,
                        customerAccessToken: token
                    };
                    return [4 /*yield*/, context.client.graphQLClient.send(customerAddressUpdateMutation(context), data).then(function (_a) {
                            var model = _a.model;
                            if (model) {
                                return model.customerAddressUpdate;
                            }
                        })];
                case 1:
                    updatedAddressId = _a.sent();
                    return [2 /*return*/, updatedAddressId];
            }
        });
    });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function checkOut(context, checkoutId, customQuery) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, context.client.checkout.fetch(checkoutId)];
        });
    });
}

var CustomClient = require('shopify-buy/index.unoptimized.umd.min.js');
var defaultSettings = {};
var cookies = {
    cartCookieName: 'vsf-cart'
};
var onCreate = function (settings) {
    return ({
        config: __assign(__assign({}, defaultSettings), settings),
        client: CustomClient.buildClient(settings.api),
        cookies: (settings.api).cookies || cookies
    });
};
var createApiClient = core.apiClientFactory({
    onCreate: onCreate,
    api: {
        getProduct: getProduct,
        getCategory: getCategory,
        getBlogPosts: getBlogPosts,
        getPages: getPages,
        editProfile: editProfile,
        addToCart: addToCart,
        updateCart: updateCart,
        removeFromCart: removeFromCart,
        signUp: signUp,
        signIn: signIn,
        signOut: signOut,
        changePassword: changePassword,
        forgotPassword: forgotPassword,
        fetchCustomer: fetchCustomer,
        fetchOrders: fetchOrders,
        fetchAddresses: fetchAddresses,
        deleteAddress: deleteAddress,
        addAddress: addAddress,
        updateAddress: updateAddress,
        createCart: createCart,
        checkOut: checkOut,
        cookies: cookies
    }
}).createApiClient;

exports.createApiClient = createApiClient;
